js/source/tile_pyramid.js

```diff
@@ -16,24 +17,65 @@ module.exports = TilePyramid;
  * data.
  *
  * @param {Object} options
- * @param {number} options.tileSize
- * @param {number} options.minzoom
- * @param {number} options.maxzoom
  * @private
  */
-function TilePyramid(options) {
-    this.tileSize = options.tileSize;
-    this.minzoom = options.minzoom;
-    this.maxzoom = options.maxzoom;
-    this.roundZoom = options.roundZoom;
-    this.reparseOverscaled = options.reparseOverscaled;
-
-    this._load = options.load;
-    this._abort = options.abort;
-    this._unload = options.unload;
-    this._add = options.add;
-    this._remove = options.remove;
-    this._redoPlacement = options.redoPlacement;
+function TilePyramid(id, options, dispatcher) {
+    this.id = id;
+    this.dispatcher = dispatcher;
+
+    var onChange = function () {
+        this.reload();
+        if (this.transform) {
+            this.update(this.transform, this.map && this.map.style.rasterFadeDuration);
+        }
+    }.bind(this);
+
+    // TODO: this is an ugly consequence of inverting tilepyramid / source
+    // dependency and making source creation async: we use a dumb, generic
+    // version of serialize() until the source is loaded at which point we
+    // switch to deferring to the source's method.
+    // Options:
+    // 1. Make Source.create synchronous.  Drawback: would require adding `loaded()` or `onLoad`  to the Source interface.
+    // 2. Remove Source:serialize() entirely, and instead let TilePyramid:serialize() always jut return the options object it was created with.
+    // 3. Leave the below wart as-is.
+    // I prefer option 2 if it's workable.
```

Comment:
By the time we add an `onLoad` parameter, it might be smart to require sources to mixin `Evented`. It's slightly more work for 3rd parties, I admit, but it helps us keep our architecture clean and consistent.